mtcpclient/groupofconnections.go

```diff
@@ -46,6 +55,59 @@ func (gc GroupOfConnections) AtLeastOneConnectionInError() bool {
 	return gc.isIn(tcpclient.ConnectionError)
 }
 
-func (gc GroupOfConnections) AtLeastOneConnectionEstablished() bool {
-	return gc.isIn(tcpclient.ConnectionEstablished)
+func (gc GroupOfConnections) AtLeastOneConnectionOK() bool {
+	return gc.isIn(tcpclient.ConnectionEstablished) || gc.isIn(tcpclient.ConnectionClosed)
+}
+
+func appendConnections(previous_connections []tcpclient.Connection, new_connections ...tcpclient.Connection) []tcpclient.Connection {
+	// TODO implement as a method using as a reference
+	m := len(previous_connections)
+	n := m + len(new_connections)
+	if n > cap(previous_connections) { // if necessary, reallocate
+		// allocate double what's needed, for future growth.
+		newSlice := make([]tcpclient.Connection, (n+1)*2)
+		copy(newSlice, previous_connections)
+		previous_connections = newSlice
+	}
+	previous_connections = previous_connections[0:n]
+	copy(previous_connections[m:n], new_connections)
+	return previous_connections
+}
+
+func (gc GroupOfConnections) getFilteredListByStatus(status []tcpclient.ConnectionStatus) (filteredConnections []tcpclient.Connection) {
+	for _, connection := range gc.connections {
+		for _, s := range status {
+			if connection.GetConnectionStatus() == s {
+				filteredConnections = appendConnections(filteredConnections, connection)
+			}
+		}
+	}
+	return filteredConnections
+}
+
+func (gc GroupOfConnections) pingStyleReport(status tcpclient.ConnectionStatus) (output string) {
+	var introduction string
+	var filteredConnections GroupOfConnections
+	var mr metricsCollectionStats
+
+	switch status {
+	case tcpclient.ConnectionEstablished:
+		filteredConnections.connections = gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionEstablished, tcpclient.ConnectionClosed})
+		mr = filteredConnections.calculateMetricsReport()
+		introduction = "Response time stats for " + strconv.Itoa(mr.numberOfConnections) +
+			" successful connections min/avg/max/dev = "
+
+	case tcpclient.ConnectionError:
+		filteredConnections.connections = gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionError})
```

Comment:
status is equal to tcpclient.ConnectionError, right? If you keep this as it is (see previous comment), you may be able to remove two lines of duplication between this and the previous case block...